From c60ac5e5d199319fcc596ded5da96f3aaab5ca38 Mon Sep 17 00:00:00 2001
From: Brad Larson <bradley.larson@amd.com>
Date: Wed, 2 Apr 2025 11:04:12 -0700
Subject: [PATCH 03/17] hwmon: (pmbus/tps53679) Add support for tps53659

Signed-off-by: Brad Larson <bradley.larson@amd.com>
---
 drivers/hwmon/pmbus/tps53679.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/hwmon/pmbus/tps53679.c b/drivers/hwmon/pmbus/tps53679.c
index 5c9466244d70..61cba9369e89 100644
--- a/drivers/hwmon/pmbus/tps53679.c
+++ b/drivers/hwmon/pmbus/tps53679.c
@@ -16,7 +16,7 @@
 #include "pmbus.h"
 
 enum chips {
-	tps53647, tps53667, tps53676, tps53679, tps53681, tps53688
+	tps53647, tps53659, tps53667, tps53676, tps53679, tps53681, tps53688
 };
 
 #define TPS53647_PAGE_NUM		1
@@ -252,6 +252,11 @@ static int tps53679_probe(struct i2c_client *client)
 	case tps53676:
 		info->identify = tps53676_identify;
 		break;
+	case tps53659:
+		info->pages = TPS53679_PAGE_NUM;
+		info->identify = tps53679_identify;
+		info->func[0] |= PMBUS_HAVE_PIN;
+		break;
 	case tps53679:
 	case tps53688:
 		info->pages = TPS53679_PAGE_NUM;
@@ -273,6 +278,7 @@ static int tps53679_probe(struct i2c_client *client)
 static const struct i2c_device_id tps53679_id[] = {
 	{"bmr474", tps53676},
 	{"tps53647", tps53647},
+	{"tps53659", tps53659},
 	{"tps53667", tps53667},
 	{"tps53676", tps53676},
 	{"tps53679", tps53679},
@@ -285,6 +291,7 @@ MODULE_DEVICE_TABLE(i2c, tps53679_id);
 
 static const struct of_device_id __maybe_unused tps53679_of_match[] = {
 	{.compatible = "ti,tps53647", .data = (void *)tps53647},
+	{.compatible = "ti,tps53659", .data = (void *)tps53659},
 	{.compatible = "ti,tps53667", .data = (void *)tps53667},
 	{.compatible = "ti,tps53676", .data = (void *)tps53676},
 	{.compatible = "ti,tps53679", .data = (void *)tps53679},
-- 
2.25.1

